aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -128,6 +163,120 @@ struct OuterSumCastLoadPolicy<scalar_t, scalar_t>:
     LoadPolicy<scalar_t> {
 };
 
+/* To implement nansum, augment the load operation to mask out nans before
+ * entering the normal sum loop.
+ */
+
+template <typename scalar_t>
+struct NanSumLoadPolicy {
+  static constexpr int64_t memsize() {
+    return sizeof(scalar_t);
+  }
+
+  static scalar_t load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto val = LoadPolicy<scalar_t>::load(data, stride, index);
+    return at::_isnan(val) ? scalar_t(0) : val;
+  }
+};
+
+template <typename scalar_t>
+struct NanSumLoadPolicy<Vectorized<scalar_t>> {
+  using vec_t = Vectorized<scalar_t>;
+
+  static constexpr int64_t memsize() {
+    return LoadPolicy<vec_t>::memsize();
+  }
+
+  static vec_t load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto val = LoadPolicy<vec_t>::load(data, stride, index);
+    return vec_t::blendv(val, vec_t(0), val.isnan());
+  }
+};
+
+template <typename scalar_t, typename acc_t>
+struct NanSumCastLoadPolicy {
+  static constexpr int64_t memsize() {
+    return sizeof(scalar_t);
+  }
+
+  static acc_t load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    using vec_t = Vectorized<scalar_t>;
+    auto val = CastLoadPolicy<scalar_t, acc_t>::load(data, stride, index);
+    return at::_isnan(val) ? acc_t(0) : val;
+  }
+};
+
+template <typename vec_t, typename vacc_t>
+struct InnerNanSumCastLoadPolicy {
```

Comment:
The key difference is the _isnan checks in the inner loop.

Reply:
Yeah, but it could be a lambda arg that would be 
```
[](scalar_t val){return !at::_isnan(val);}
```
for the nan case and 
```
[](scalar_t val){return true;}
```
for the regular sum case?

Reply:
@ngimel I've implemented a slight variation on this.